REGRESSION (299407@main): Google AI page renders incorrectly

weinig opened the pull request:
#### 9b938cd84e6c5f652b5af964839d613b879cf553
<pre>
REGRESSION (299407@main): Google AI page renders incorrectly
<a href="https://bugs.webkit.org/show_bug.cgi?id=301429">https://bugs.webkit.org/show_bug.cgi?id=301429</a>

Reviewed by Antti Koivisto.

Bandaid fix to allow custom property animations that update background-image
or mask-image to work by updating the FillLayers list rather than replacing
it. A better long term fix is to use CoordinatedValueList, like `animation`
and `transition`. That is tracked in <a href="https://webkit.org/b/301581.">https://webkit.org/b/301581.</a>

Test: fast/backgrounds/background-image-style-update-by-custom-property-animation.html
* LayoutTests/fast/backgrounds/background-image-style-update-by-custom-property-animation-expected.txt: Added.
* LayoutTests/fast/backgrounds/background-image-style-update-by-custom-property-animation.html: Added.
* Source/WebCore/css/scripts/process-css-properties.py:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/values/backgrounds/StyleFillLayers.h:

Canonical link: <a href="https://commits.webkit.org/302305@main">https://commits.webkit.org/302305@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4f296f40a4064b035eebdab2427315103ef81545

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/128666 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/931 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/39500 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136053 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/80066 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/130538 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/880 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/807 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97947 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/65865 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/131614 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/653 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115275 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78565 "Passed tests") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/588 "Build is in progress. Recent messages:OS: Tahoe (26.0), Xcode: 26.0.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Passed layout tests; Running layout-tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/33385 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79334 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109035 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/33864 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/138508 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/756 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/717 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/106483 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/805 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111614 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106306 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27074 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/635 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30144 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/53136 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/818 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/64069 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/682 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/737 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/765 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Ahmad-S792 commented on LayoutTests/fast/backgrounds/background-image-style-update-by-custom-property-animation.html:
```diff
@@ -0,0 +1,28 @@
+<script src="../../resources/testharness.js"></script>
```
Is it intentionally in quirk mode?

weinig replied:
No. It’s the same in both strict and quirks mode.

anttijk commented on Source/WebCore/style/StyleBuilderCustom.h:
```diff
@@ -280,23 +280,67 @@ void applyInheritPrimaryFillLayerProperty(BuilderState& builderState)
 
-template<auto layersSetter, auto converter, typename Layers>
+template<auto layersMutableGetter, auto layersSetter, auto setter, auto getter, auto initial, auto converter, typename Layers>
 void applyValuePrimaryFillLayerProperty(BuilderState& builderState, CSSValue& value)
```
do initial/inherit have the same issue?

weinig replied:
I don't think so. I haven't been able to find a test case that fails in the same way due to the use of custom properties here.

weinig committed:
REGRESSION (299407@main): Google AI page renders incorrectly
https://bugs.webkit.org/show_bug.cgi?id=301429

Reviewed by Antti Koivisto.

Bandaid fix to allow custom property animations that update background-image
or mask-image to work by updating the FillLayers list rather than replacing
it. A better long term fix is to use CoordinatedValueList, like `animation`
and `transition`. That is tracked in https://webkit.org/b/301581.

Test: fast/backgrounds/background-image-style-update-by-custom-property-animation.html
* LayoutTests/fast/backgrounds/background-image-style-update-by-custom-property-animation-expected.txt: Added.
* LayoutTests/fast/backgrounds/background-image-style-update-by-custom-property-animation.html: Added.
* Source/WebCore/css/scripts/process-css-properties.py:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/values/backgrounds/StyleFillLayers.h:

Canonical link: https://commits.webkit.org/302305@main
